test/generators/html/gen_html/gen_html.ml

```diff
@@ -0,0 +1,103 @@
+open Sexplib0.Sexp
+
+type sexp = t = Atom of string | List of sexp list
+
+let read_lines ic =
+  let lines = ref [] in
+  try
+    while true do
+      lines := input_line ic :: !lines
+    done;
+    assert false
+  with End_of_file -> List.rev !lines
+
+let lines_of_command prog args =
+  let cmd = String.concat " " (prog :: List.map Filename.quote args) in
+  let inp = Unix.open_process_in cmd in
+  let lines = read_lines inp in
+  let status =
+    match Unix.close_process_in inp with
+    | Unix.WEXITED n -> `Exit n
+    | _ -> `Signal
+  in
+  (lines, status)
+
+let expected_html_targets odocl =
+  lines_of_command "odoc html-targets" [ "-o"; "html.gen"; odocl ]
+
+let tweak_html_target target =
+  match String.split_on_char '/' target with
+  | _ :: _ :: rest -> String.concat "." rest
+  | _ -> assert false
+
+let die s =
+  prerr_endline s;
+  exit 1
+
+let html_gen_targets targets status =
+  match status with
+  | `Exit 0 ->
+      List.map
+        (fun t ->
+          List
+            [
+              Atom "with-stdout-to";
+              Atom (tweak_html_target t ^ ".gen");
+              List [ Atom "progn"; List [ Atom "system"; Atom ("cat " ^ t) ] ];
+            ])
+        targets
+  | _ ->
+      [
+        die
+          (Filename.quote "odoc html-generate"
+          ^ "command execution not successful!");
+      ]
+
+let target_rule odocl targets status =
+  List
+    [
+      Atom "rule";
+      List
+        [
+          Atom "action";
+          List
+            (Atom "progn"
+             ::
+             List
+               [
+                 Atom "run";
+                 Atom "odoc";
+                 Atom "html-generate";
+                 Atom "--indent";
+                 Atom "-o";
+                 Atom "html.gen";
+                 Atom ("%{dep:" ^ odocl ^ "}");
+               ]
+             :: html_gen_targets targets status);
+        ];
+    ]
+
+let diff_rule t =
+  List
+    [
+      Atom "rule";
+      List [ Atom "alias"; Atom "runtest" ];
+      List [ Atom "action"; List [ Atom "diff"; Atom t; Atom (t ^ ".gen") ] ];
+    ]
+
+let diff_rules targets =
+  List.iter
+    (fun t -> Printf.printf "%s\n\n" (to_string_hum (diff_rule t)))
+    targets
+
+let gen_html_rule filenames =
+  filenames
+  |> List.iter (fun odocl ->
+         let targets, status = expected_html_targets odocl in
+         Printf.printf "%s\n\n"
+           (to_string_hum (target_rule odocl targets status));
+         diff_rules (List.map tweak_html_target targets))
+
+let () =
+  let files = List.tl (Array.to_list Sys.argv) in
```

Comment:
I think we should use fpath in this program too. `files` should be converted to fpath as soon as possible. `tweak_html_target` will use `Fpath.segs` instead of `String.split_on_char`.

Reply:
done